Docs for logging, events, and diagnostics (#2787)
dotnet/EntityFramework.Docs: entity-framework/core/miscellaneous/events/index.md

ms.date: 10/01/2020
uid: core/miscellaneous/events/index
---

# Overview of logging and interception

> Entity Framework Core (EF Core) contains several mechanisms for generating logs, responding to events, and obtaining diagnostics. Each of these are tailored to different situations, and it is important to select the best mechanism for the task in hand, even when multiple mechanisms could work. For example, a database interceptor could be used to log SQL, but this is better handled by one of the mechanisms tailored to logging. This page presents an overview of each of these mechanisms and describes when each should be used.

Reviewer: ```suggestion
Entity Framework Core (EF Core) contains several mechanisms for generating logs, responding to events, and obtaining diagnostics. Each of these is tailored to different situations, and it is important to select the best mechanism for the task in hand, even when multiple mechanisms could work. For example, a database interceptor could be used to log SQL, but this is better handled by one of the mechanisms tailored to logging. This page presents an overview of each of these mechanisms and describes when each should be used.
```

## Quick reference

The table below provides a quick reference for the differences between the mechanisms described here.

| Mechanism |  Async | Scope | Registered | Intended use
|:----------|--------|-------|------------|-------------
| Simple Logging | No | Per context | Context configuration | Development-time logging
| Microsoft.Extensions.Logging | No | Per context* | D.I. or context configuration | Production logging
| Events | No | Per context | Any time | Reacting to EF events
| Interceptors | Yes | Per context | Context configuration | Manipulating EF operations
| Diagnostics listeners | No | Process | Globally | Application diagnostics

> *Typically `Microsoft.Extensions.Logging` is configured per-application in D.I. However, at the EF level, each context _can_ be configured with a different logger if needed.

Reviewer 2: ```suggestion
*Typically `Microsoft.Extensions.Logging` is configured per-application via dependency injection (DI). However, at the EF level, each context _can_ be configured with a different logger if needed.
```

This concept is similar to <xref:System.Data.Entity.Database.Log?displayProperty=nameWithType> in EF6.

> See [Simple logging](xref:core/miscellaneous/events/simple-logging) for more information.

Reviewer 2: Above: "EF Core logs can be accessed from any type of application through use of" should be "EF Core logs can be accessed from any type of application through *the* use of"

## Microsoft.Extensions.Logging

> [Microsoft.Extensions.Logging](/dotnet/core/extensions/logging) is an extensible logging mechanism with plug-in providers for many common logging systems.
> 
> EF Core fully integrates with `Microsoft.Extensions.Logging` and this form of logging is used by default for ASP.NET Core applications.

Reviewer 2: ```suggestion
[Microsoft.Extensions.Logging](/dotnet/core/extensions/logging) is an extensible logging mechanism with plug-in providers for many common logging systems. EF Core fully integrates with `Microsoft.Extensions.Logging` and this form of logging is used by default for ASP.NET Core applications.
```

See [Using Microsoft.Extensions.Logging in EF Core](xref:core/miscellaneous/events/extensions-logging) for more information.

## Events

> [!NOTE]
> Additional events were added in EF Core 5.0.

> EF Core exposes [.NET events](/dotnet/standard/events/) to act as callbacks when certain things happen in the EF Core code. Events are simpler than interceptors and allow more flexible registration. However, they are sync only and so cannot perform non-blocking async I/O.

Reviewer 2: ```suggestion
EF Core exposes [.NET events](/dotnet/standard/events/) to act as callbacks when certain things happen in the EF Core code. Events are simpler than interceptors and allow more flexible registration, but are sync only and so cannot perform non-blocking async I/O.
```
